File: EPI/10.9/swift/10.9_test.swift
// swiftlint:disable variable_name

//     0
//    / \
//   1   2
//  / \   \
// 3   4   5
//    /
//   6

let t = Node(
    value: 0,
    left: Node(
        value: 1,
        left: Node(value: 3),
        right: Node(
            value: 4,
            left: Node(value: 6)
        )
    ),
    right: Node(
        value: 2,
        right: Node(value: 5)
    )
)

assert(t.inorder() == [3, 1, 6, 4, 0, 2, 5])

File: EPI/10.9/typescript/10.9_test.ts
import {Node, traverseInOrder} from './10.9';

declare function require(name: string): any;
const assert = require('assert');

interface NumberedNode extends Node {
    value: number;
}

//     0
//    / \
//   1   2
//  / \   \
// 3   4   5
//    /
//   6

const [zero, one, two, three, four, five, six]: NumberedNode[] =
    Array.from({length: 7}, (_, i) => ({value: i}));

zero.left = one;
zero.right = two;

one.parent = zero;
one.left = three;
one.right = four;

two.parent = zero;
two.right = five;

three.parent = one;

four.parent = one;
four.left = six;

five.parent = two;

six.parent = four;

const actual: number[] = [];
traverseInOrder(zero, node => actual.push(node.value));

assert.deepStrictEqual(actual, [3, 1, 6, 4, 0, 2, 5]);
